### ISSUE TITLE
 FormatException with Byte data on topic

### ISSUE THREAD

**Issue by dellorogiulio** [ra1u/redis-dart#56]
Hi everyone, 
I'm doing some tests with this package and I found a problem when interfacing with a redis on which topic messages are messagepack-packed data [official site](https://msgpack.org/). 
Until on topic there are strings only, this package has no problem using the following code:
```dart
_subscriber!.getStream().listen((event) {
   print(event);
}
```
however, when I try to receive an "encoded" message like this:
![image](https://user-images.githubusercontent.com/59453641/152311711-ea52431b-1c80-46bf-82e2-79e9ed79564b.png)

which is shown as:
- `"\x91\x00"` on the _redis-cli_ console tool on windows
- `<Buffer 91 00>` on a node.js application 

I get
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: FormatException: Unexpected extension byte (at offset 0)
#0      _Utf8Decoder.convertSingle (dart:convert-patch/convert_patch.dart:1789:7)
#1      Utf8Decoder.convert (dart:convert/utf.dart:318:42)
#2      Utf8Codec.decode (dart:convert/utf.dart:63:20)
#3      RedisParser.parseBulk.<anonymous closure>.<anonymous closure>
package:redis/redisparser.dart:95
#4      _rootRunUnary (dart:async/zone.dart:1436:47)
#5      _CustomZone.runUnary (dart:async/zone.dart:1335:19)
<asynchronous suspension>
#6      _CustomZone.bindUnaryCallbackGuarded.<anonymous closure> (dart:async/zone.dart)
<asynchronous suspension>
```

*before* the
```dart
subscriber!.getStream().listen
```
callback is called.

How can we fix this?

Thanks in advance

**Comment by ra1u** [comment]
I think issue is that we transform received bulk data as strings:

https://github.com/ra1u/redis-dart/blob/master/lib/redisparser.dart#L95

For quick hack you should be fine if you throw out conversion UTF8.decode, but for backward compatibility i will have to came out with better idea.

**Comment by dellorogiulio** [comment]
Hi @ra1u ,
thanks for your response. I try and it works! I mean, now `event`  in 
```dart
_subscriber!.getStream().listen((event) {
   print(event);
}
```
is a List of List (of course) 
![image](https://user-images.githubusercontent.com/59453641/152319815-cb33a58f-7b3c-497a-a5c4-46dde45f4667.png)
so I need to decode both `event[0]`  and `event[1]`

```dart
const UTF8 = Utf8Codec();
String action = UTF8.decode(event[0]);
if (action != 'message') {
return;
}
String topicName = UTF8.decode(event[1]);
```
which doesn't sound so bad to me. However, a friendly interface is always a good thing.

As implementation idea, you could mimic [node ioredis package](https://github.com/luin/ioredis#pubsub)  which allows user to subscribe on 'message' (consider a string) or 'messageBuffer' which is instead a raw buffer.
Obviously this need some complexity from your side since you should have to wrap dart Stream and dispatch different callback for different events (now every event has the same callback and then user had to handle what it need), so that the user can register a callback to the 'subscribe' action, to the 'message', 'messageBuffer' etc. 

Note: in ioredis, when you subscribe to 'messageBuffer', topic name is raw buffer too.

What do you think about that?

Thanks again!

**Comment by dellorogiulio** [comment]
@ra1u  I add another issue here: 
If I try to publish a simple string using:
```dart
await _redis!.send_object(["PUBLISH", "topic_2", "message"]);
```

however I receive:
```
[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: type 'RedisError' is not a subtype of type 'List<dynamic>'
#0      new PubSub.<anonymous closure>.<anonymous closure>.<anonymous closure>
package:redis/pubsub.dart:20
#1      _rootRunUnary (dart:async/zone.dart:1436:47)
#2      _CustomZone.runUnary (dart:async/zone.dart:1335:19)
<asynchronous suspension>
#3      _CustomZone.bindUnaryCallbackGuarded.<anonymous closure> (dart:async/zone.dart)
<asynchronous suspension>
```

Just to completeness: what if I need to publish raw data? Something like a UInt8List as third argument?
```
await _redis!.send_object(["PUBLISH", "topic_2", [1, 2, 3, 4]]);
``` 

Thanks in advance

**Comment by ra1u** [comment]
For that, take a peek in https://github.com/ra1u/redis-dart/blob/master/lib/redisserialise.dart
There is class [`RedisBulk`](https://github.com/ra1u/redis-dart/blob/master/lib/redisserialise.dart#L13) that you can use for this.
(just push your data into this class).

**Comment by dellorogiulio** [comment]
Sorry but I do not understand, 
`await _redis!.send_object(["PUBLISH", "topic_2", "message"]);` should work out-of-the-box as reported in your documentation.
Am I missing something?

**Comment by ra1u** [comment]
For your case that should be 

~~~~~Dart
await _redis!.send_object(["PUBLISH", "topic_2", RedisBulk([1, 2, 3, 4])]);
~~~~~

**Comment by ra1u** [comment]
We have now available support for binary data on master. Idea is that parser is embedded in class Command.

~~~~Dart
final conn = RedisConnection();
Command _cmd = await conn.connect('localhost',6379);
Command cmd_binary = Command.from(_cmd).setParser(RedisParserBulkBinary()); 

List<int> data = [1,2,3,4,5,6,7,8,9];    
await cmd_binary.send_object(["SET", "mykey", RedisBulk(data)]); 
List<int> rsp = cmd_binary.send_object(["GET", key]);
~~~~

Due to minor backward incompatibility it is expected to be released as v3.0 in near future.

**Comment by dellorogiulio** [comment]
Thanks!
